src/mscorlib/shared/System/Text/StringBuilder.cs

```diff
@@ -790,6 +790,123 @@ public StringBuilder Append(string value, int startIndex, int count)
             }
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            if (value != null && value.Length != 0)
+            {
+                if (value != this)
+                    AppendCore(value, 0, value.Length);
+                else
+                    Append(value.ToString());
+            }
+            return this;
+        }
+
+        public StringBuilder Append(StringBuilder value, int startIndex, int count)
+        {
+            if (startIndex < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_GenericPositive);
+            }
+
+            if (value == null)
+            {
+                if (startIndex == 0 && count == 0)
+                {
+                    return this;
+                }
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            if (count > value.Length - startIndex)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_Index);
+            }
+
+            if (count > 0)
+            {
+                if (value != this)
+                    AppendCore(value, startIndex, count);
+                else
+                    Append(value.ToString(startIndex, count));
+            }
+            return this;
+        }
+
+        private void AppendCore(StringBuilder value, int startIndex, int count)
+        {
+            int newLength = Length + count;
+            if ((uint)newLength > (uint)m_MaxCapacity)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_SmallMaxCapacity);
+            }
+
+            while (count > 0)
+            {
+                int length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
+                if (length == 0)
+                {
+                    ExpandByABlock(count);
+                    length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
+                }
+
+                StringBuilder chunk = value;
+                int sourceEndIndex = startIndex + length;
+                int curDestIndex = length;
+                unsafe
```

Comment:
This unsafe block is identical to the same one in ToString(). I would extract a method they can share
`private static unsafe char* CopyToBuffer(StringBuilder chunk, char* destinationPtr, int startIndex, int length)`

Reply:
I think the body of that should be able to itself reuse the exissting
`private static unsafe void ThreadSafeCopy(char* sourcePtr, char[] destination, int destinationIndex, int count)`

Reply:
Should I change this for  ToString() also ?

Reply:
@jkotas is reuse here better than avoiding a function call?

Reply:
All these methods and loops with pointers passed around are unnecessary security risk. 

I would make this method 100% safe and just call `Append(char[] value, int startIndex, int charCount)` in a loop here.

If you want to go extra mile and save some of the unnecessary argument validation overhead, change the core append method to `Append(ReadOnlySpan<char>)` and make other methods including this one call it.

Reply:
BTW: `Append(ReadOnlySpan<char>)` would be a good public API to add - I expect it will be no-brainer to approve.

Reply:
> @jkotas is reuse here better than avoiding a function call?

I see what you meant now. My suggestion with Append would not work/is unrelated.

I would optimize for minimizing/eliminating unsafe code. The StringBuilder code is not easy to reason about. We had number of security issues in it over the years. Reducing code duplication would be secondary for me.

Reply:
Maybe something like this?

```
```
int newLength = Length + count;
if ((uint)newLength > (uint)m_MaxCapacity)
{
    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
}

while (count > 0)
{
    int length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
    if (length == 0)
    {
        ExpandByABlock(count);
        length = Math.Min(m_ChunkChars.Length - m_ChunkLength, count);
    }

    value.CopyTo(startIndex, new Span<char>(m_ChunkChars, m_ChunkLength, length), count);

    startIndex += length;
    count -= length;
}
```

Reply:
CopyTo has same loop as what  you have here. Calling CopyTo from here avoids the code duplication - makes the code smaller I think.

There is probably a cleanup that can be done around ThreadSafeCopy - delete the one that take pointer. I would recommend doing that as separate PR if you are interested in it.

Reply:
I can do the cleanup regarding the ThreadSafeCopy and toString(int,int) in another PR

Reply:
@jkotas which one of these is preferable for toString(int,int). We are using extra memory in first one but there is no unsafe code in this.
```
AssertInvariants();
char[] destination = new char[length];
this.CopyTo(startIndex, new Span<char>(destination), length);
return new string(destination);
```
```
AssertInvariants();
 string result = string.FastAllocateString(length);
  unsafe
   {
         fixed (char* destinationPtr = result)
    {
           this.CopyTo(startIndex, new Span<char>(destinationPtr, length), length);
            return result;
       }
}
```

Reply:
You certainly do not want to allocate temporary `char[]`.

You second option would be ok.

If you wanted to avoid unsafe code altogether, you would write it as:
```
return string.Create(length,
   (Source: this, StartIndex: startIndex),
   (span, state) => state.Source.CopyTo(state.StartIndex, span, span.Length));
```

The C# looks better, but it is a bit slower and quite a bit bigger code. I think I like your second option with a tiny bit of unsafe code better in this particular case.